docs/running_lbann.rst

```diff
@@ -1,96 +1,434 @@
 .. role:: bash(code)
           :language: bash
+.. role:: python(code)
+          :language: python
 
-====================
+============================================================
 Running LBANN
-====================
+============================================================
 
-The basic template for running LBANN is
+------------------------------------------------
+Anatomy of an LBANN experiment
+------------------------------------------------
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Parallelism
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+LBANN is run under the `MPI
+<https://en.wikipedia.org/wiki/Message_Passing_Interface>`_ paradigm,
+i.e. with multiple processes that communicate with message
+passing. These processes are subdivided into "trainers." Conceptually,
+a trainer owns parallel objects, like models and data readers, and
+generally operates independently of other trainers.
+
+Comments:
+
++ LBANN targets HPC systems with homogeneous compute nodes and GPU
+  accelerators, which motivates some simplifying assumptions:
+
+  - All trainers have the same number of processes.
+
+  - Each MPI process corresponds to one GPU.
+
++ Processors are block assigned to trainers based on MPI rank.
+
+  - In order to minimize the cost of intra-trainer communication, make
+    sure to map processes to the hardware and network
+    topologies. Typically, this just means choosing a sensible number
+    of processes per trainer, e.g. a multiple of the number of GPUs
+    per compute node.
+
++ Generally, increasing the number of processes per trainer will
+  accelerate computation but require more intra-trainer
+  communication. There is typically a sweet spot where run time is
+  minimized, but it is complicated and sensitive to the type of
+  computational operations, the amount of work, the hardware and
+  network properties, and the communication algorithms.
+
+  - Rule-of-thumb: Configure experiments so that the bulk of run time
+    is taken by compute-bound operations (e.g. convolution or matrix
+    multiplication) and so that each process has enough work to
+    achieve a large fraction of peak performance (e.g. by making the
+    mini-batch size sufficiently large).
+
++ Most HPC systems are managed with job schedulers like `Slurm
```

Comment:
It might be good to include LSF commands in addition to the Slurm ones.

Reply:
I agree it might be good to include LSF commands, here is one I use for interactive run:
bsub -q pdebug -nnodes 16 -Is -W 30 -G lbpm /usr/bin/bash  (16 nodes for 30 minutes, debug queue, lbpm account etc). Refer to corresponding LC resources for LSF